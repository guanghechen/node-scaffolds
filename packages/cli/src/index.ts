export * from './util/ansi'
export * from './util/command-exists'
export * from './util/config'
export * from './util/merge'
export * from './util/option'
export * from './util/stdin'
export * from './types'
