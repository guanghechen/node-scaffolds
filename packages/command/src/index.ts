export * from './command-top'
export * from './command'
export * from './definition/command'
export * from './definition/option'
export * from './util'
