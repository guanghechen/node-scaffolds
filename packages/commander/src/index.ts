export * from './command'
export * from './main'
export * from './sub'
export * from './top'
export * from './util'

export * from '@guanghechen/cli'
