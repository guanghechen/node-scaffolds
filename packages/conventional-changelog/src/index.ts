import { parserOpts, recommendedBumpOpts, writerOpts as writerOptsPromise } from './opts'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
async function buildPresetConfig() {
  const writerOpts = await writerOptsPromise
  const conventionalChangelog = { parserOpts, writerOpts }
  return {
    conventionalChangelog,
    parserOpts,
    recommendedBumpOpts,
    writerOpts,
  }
}

const presetConfig: any = buildPresetConfig()
export default presetConfig
