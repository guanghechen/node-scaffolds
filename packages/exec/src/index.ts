import type { IReporter } from '@guanghechen/reporter.types'
import { spawn } from 'node:child_process'

export interface ISafeExecParams {
  from: string
  cmd: string
  args: string[]
  cwd?: string
  env?: Record<string, string | undefined>
  stdio?: 'pipe' | 'overlapped' | 'inherit' | 'ignore'
  encoding?: BufferEncoding
  reporter?: IReporter
}

export interface ISafeExecResult {
  stdout: string
}

export async function safeExec(params: ISafeExecParams): Promise<ISafeExecResult> {
  const { from, cmd, args, cwd, stdio, encoding = 'utf8', reporter } = params
  const env: Record<string, string | undefined> = { ...process.env, ...params.env }
  reporter?.debug(
    '[safeExec] {}: cmd: {}, args: {}, cwd: {}, env: {}, encoding: {}',
    from,
    cmd,
    args,
    cwd,
    env,
    encoding,
  )

  try {
    const stdout: string = await new Promise<string>((resolve, reject) => {
      const command = spawn(cmd, args, { cwd, env, stdio })

      let stdoutData: string = ''
      let stderrData: string = ''

      command.stdout!.on('data', data => {
        stdoutData += data.toString(encoding)
      })

      command.stderr!.on('data', data => {
        stderrData += data.toString(encoding)
      })

      command.on('close', code => {
        if (code === 0) resolve(stdoutData)
        else reject(stderrData)
      })
    })

    const result: ISafeExecResult = { stdout }
    return result
  } catch (error) {
    reporter?.error?.('[safeExec] failed to run {}, (args: {}, params: {})', cmd, args, params)
    throw error
  }
}
