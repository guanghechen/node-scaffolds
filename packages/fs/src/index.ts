export * from './collect'
export * from './enhanced'
export * from './is'
