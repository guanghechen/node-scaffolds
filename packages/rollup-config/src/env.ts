import type { IEnv } from './types'

export interface IRawRollupConfigEnv {
  /**
   * Whether if generate sourcemaps.
   * @default true
   */
  sourcemap?: boolean
  /**
   * Whether if make all dependencies external.
   * @default true
   */
  external?: boolean
}

export function resolveRollupConfigEnv(rawEnv: IRawRollupConfigEnv): IEnv {
  const forceSourceMap: boolean | undefined = convertToBoolean(
    process.env['ROLLUP_SHOULD_SOURCEMAP'],
  )
  const forceExternalAll: boolean | undefined = convertToBoolean(
    process.env['ROLLUP_EXTERNAL_ALL_DEPENDENCIES'],
  )
  const env: IEnv = {
    sourcemap: forceSourceMap ?? rawEnv.sourcemap ?? true,
    externalAll: forceExternalAll ?? rawEnv.external ?? true,
  }
  return env
}

function convertToBoolean(v?: unknown): boolean | undefined {
  if (typeof v === 'boolean') return v
  if (typeof v === 'string') {
    const t = v.toLowerCase()
    if (t === 'false') return false
    if (t === 'true') return true
  }
  return undefined
}
