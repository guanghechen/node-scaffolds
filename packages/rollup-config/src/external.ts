import { DependencyCategory } from './types'
import type { IEnv, IManifestWithDependencies } from './types'

export const builtinExternalSet: ReadonlySet<string> = new Set<string>([
  ...[
    'assert',
    'async_hooks',
    'buffer',
    'child_process',
    'cluster',
    'console',
    'constants',
    'crypto',
    'dgram',
    'dns',
    'domain',
    'events',
    'fs',
    'http',
    'http2',
    'https',
    'inspector',
    'module',
    'net',
    'os',
    'path',
    'perf_hooks',
    'process',
    'punycode',
    'querystring',
    'readline',
    'repl',
    'stream',
    'string_decoder',
    'timers',
    'tls',
    'trace_events',
    'tty',
    'url',
    'util',
    'v8',
    'vm',
    'wasi',
    'worker_threads',
    'zlib',
  ],
  ...['glob', 'sync'],
])

export async function resolveExternal(
  manifest: IManifestWithDependencies,
  env: IEnv,
  classifyDependency: (id: string) => DependencyCategory = () => DependencyCategory.UNKNOWN,
): Promise<(id: string) => boolean> {
  const { collectAllDependencies, getDefaultDependencyFields } = await import(
    '@guanghechen/helper-npm'
  )

  const dependencyFields = getDefaultDependencyFields()
  let dependencies: string[] = dependencyFields.reduce((acc, key) => {
    const deps = manifest[key] as Record<string, string> | string[]
    const result: string[] = Array.isArray(deps) ? deps : Object.keys(deps || {})
    return acc.concat(result)
  }, [] as string[])

  if (env.externalAll) {
    dependencies = await collectAllDependencies(null, dependencyFields, dependencies, () => true)
  }
  const externalSet = new Set(dependencies)

  const external = (id: string): boolean => {
    const classify = classifyDependency(id)
    switch (classify) {
      case DependencyCategory.BUILTIN:
        return false
      case DependencyCategory.EXTERNAL:
        return true
      case DependencyCategory.UNKNOWN:
        break
      default:
        throw new TypeError('Unknown return value from classifyDependency')
    }

    if (builtinExternalSet.has(id)) return true
    if (id.startsWith('node:')) return true

    const m = /^([.][\s\S]*|@[^/\s]+[/][^/\s]+|[^/\s]+)/.exec(id)
    if (m == null) return false
    return externalSet.has(m[1])
  }
  return external
}
