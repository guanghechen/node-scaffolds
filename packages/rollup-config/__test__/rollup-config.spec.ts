import type { IConsoleMock } from '@guanghechen/helper-jest'
import { createConsoleMock } from '@guanghechen/helper-jest'
import { desensitize } from 'jest.helper'
import fs from 'node:fs/promises'
import path from 'node:path'
import url from 'node:url'
import { rollup } from 'rollup'
import type { OutputOptions, RollupOutput } from 'rollup'
import { createRollupConfig, tsPresetConfigBuilder } from '../src'

const __dirname = path.dirname(url.fileURLToPath(import.meta.url))

// Resolve absolute dirPath of case
const resolveCaseDir = (title: string): string => path.resolve(__dirname, 'fixtures', title)

async function build(dependencies: string[]): Promise<RollupOutput[]> {
  // Create rollup config
  const configs = await createRollupConfig({
    manifest: {
      name: '@guanghechen/rollup-config',
      source: 'src/index.ts',
      main: 'lib/cjs/index.js',
      module: 'lib/esm/index.js',
      types: 'lib/types/index.d.ts',
      dependencies: dependencies,
    },
    presetConfigBuilders: [
      tsPresetConfigBuilder({
        typescriptOptions: {
          tsconfig: 'tsconfig.lib.json',
        },
      }),
    ],
  })

  const results: any[] = []
  for (const config of configs) {
    const bundle = await rollup(config)
    const outputOptions: OutputOptions[] = [config.output ?? []].flat()
    for (const outputOption of outputOptions) {
      const outputs = await bundle.generate(outputOption)
      for (const output of outputs.output as any[]) {
        const { format } = outputOption
        if (output.code != null) {
          results.push({ format, code: output.code })
        } else {
          results.push({ format, ...output })
        }
      }
    }
  }
  return results
}

afterEach(async () => {
  await fs.rm('lib', { recursive: true, force: true })
  process.chdir(__dirname)
})

describe('build', () => {
  let logMock: IConsoleMock
  beforeEach(() => {
    logMock = createConsoleMock(['warn'])
  })
  afterEach(async () => {
    logMock.restore()
  })

  it('simple', async () => {
    const caseDir = resolveCaseDir('simple')
    process.chdir(caseDir)
    const results = await build(['@guanghechen/rollup-config'])
    for (const result of results) {
      const { format, ...output } = result as any
      const data = output.code != null ? output.code : output
      expect(data).toMatchSnapshot(`rollup ${format}`)
    }

    expect(desensitize(logMock.getIndiscriminateAll())).toEqual([])
    // expect(existsSync(path.join(caseDir, 'lib/types/index.d.ts'))).toBeTruthy()
  }, 60000)
})
