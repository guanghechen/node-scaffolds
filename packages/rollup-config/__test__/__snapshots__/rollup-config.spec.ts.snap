// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build simple: rollup cjs 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('node:fs');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefault(fs);

var data = {
  name: "alice"
};
var data$1 = {
  data: data
};

function isFileSync(filepath) {
    if (!fs__default.default.existsSync(filepath))
        return false;
    return fs__default.default.statSync(filepath).isFile();
}

console.log('data:', data$1);
console.log(isFileSync(__filename));
var index = (function () { return data$1; });

exports.default = index;
exports.isFileSync = isFileSync;
//# sourceMappingURL=index.js.map
"
`;

exports[`build simple: rollup cjs 2`] = `
{
  "fileName": "index.js.map",
  "name": undefined,
  "names": [],
  "needsCodeReference": false,
  "originalFileName": null,
  "originalFileNames": [],
  "source": "{"version":3,"file":"index.js","sources":["../../src/util.ts","../../src/index.ts"],"sourcesContent":[null,null],"names":["fs","data"],"mappings":";;;;;;;;;;;;;;;;;AAQM,SAAU,UAAU,CAAE,QAAgB,EAAA;AAC1C,IAAA,IAAI,CAACA,mBAAE,CAAC,UAAU,CAAC,QAAQ,CAAC;AAAE,QAAA,OAAO,KAAK;IAC1C,OAAOA,mBAAE,CAAC,QAAQ,CAAC,QAAQ,CAAC,CAAC,MAAM,EAAE;AACvC;;ACNA,OAAO,CAAC,GAAG,CAAC,OAAO,EAAEC,MAAI,CAAC;AAC1B,OAAO,CAAC,GAAG,CAAC,UAAU,CAAC,UAAU,CAAC,CAAC;AAEnC,YAAA,CAAe,YAAM,EAAA,OAAAA,MAAI,CAAA,EAAA;;;;;"}",
  "type": "asset",
}
`;

exports[`build simple: rollup esm 1`] = `
"import fs from 'node:fs';

var data = {
  name: "alice"
};
var data$1 = {
  data: data
};

function isFileSync(filepath) {
    if (!fs.existsSync(filepath))
        return false;
    return fs.statSync(filepath).isFile();
}

console.log('data:', data$1);
console.log(isFileSync(__filename));
var index = (function () { return data$1; });

export { index as default, isFileSync };
//# sourceMappingURL=index.js.map
"
`;

exports[`build simple: rollup esm 2`] = `
{
  "fileName": "index.js.map",
  "name": undefined,
  "names": [],
  "needsCodeReference": false,
  "originalFileName": null,
  "originalFileNames": [],
  "source": "{"version":3,"file":"index.js","sources":["../../src/util.ts","../../src/index.ts"],"sourcesContent":[null,null],"names":["data"],"mappings":";;;;;;;;;AAQM,SAAU,UAAU,CAAE,QAAgB,EAAA;AAC1C,IAAA,IAAI,CAAC,EAAE,CAAC,UAAU,CAAC,QAAQ,CAAC;AAAE,QAAA,OAAO,KAAK;IAC1C,OAAO,EAAE,CAAC,QAAQ,CAAC,QAAQ,CAAC,CAAC,MAAM,EAAE;AACvC;;ACNA,OAAO,CAAC,GAAG,CAAC,OAAO,EAAEA,MAAI,CAAC;AAC1B,OAAO,CAAC,GAAG,CAAC,UAAU,CAAC,UAAU,CAAC,CAAC;AAEnC,YAAA,CAAe,YAAM,EAAA,OAAAA,MAAI,CAAA,EAAA;;;;"}",
  "type": "asset",
}
`;

exports[`build simple: rollup esm 3`] = `
"/**
 * Check if the given filepath is an existed path of file.
 *
 * @param filepath
 * @returns
 */
declare function isFileSync(filepath: string): boolean;

declare const _default: () => {
    data: {
        name: string;
    };
};

export { _default as default, isFileSync };
"
`;
